sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -419,8 +422,13 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
       }
 
       if (mode == SaveMode.Overwrite && tableExists) {
-        JdbcUtils.dropTable(conn, table)
-        tableExists = false
+        if (extraOptions.getOrElse("truncate", "false").toBoolean &&
+            JdbcUtils.isCascadingTruncateTable(url) == Some(false)) {
+          JdbcUtils.truncateTable(conn, table)
```

Comment:
Currently, it raises exceptions if one of the column names is different.
For the different column type with same column name, it works like `SaveMode.Append` operation.
I thought for the trade-off between DROP and TRUNCATE.

Let me think about the decision point.